operator/README.md

```diff
@@ -0,0 +1,179 @@
+| deviceMode | string | Device management mode to use. Supports one of `lvm` or `direct` | `lvm`
+| controllerResources | [ResourceRequirements](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#resourcerequirements-v1-core) | Describes the compute resource requirements for controller pod |
+| nodeResources | [ResourceRequirements](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#resourcerequirements-v1-core) | Describes the compute resource requirements for the pods running on node(s) |
+| nodeSelector | string map | [Labels to use for selecting Nodes](../README.md#run-pmem-csi-on-kubernetes) on which PMEM-CSI driver should run. | `{ "storage": "pmem" }`
+
+<sup>1</sup> Image versions are depend on the Kubernetes cluster version.
+
+#### DeploymentStatus
+
+A PMEM-CSI Deployment's `status` field is a `DeploymentStatus` object, which has
+a `phase` field. The phase of a Deployment is high-level summary of where the
+Deployment is in it's lifecycle.
+
+The possible `phase` values and their meaning are as below:
+
+| value | meaning |
+|---|---|---|
+| Pending | The `Deployment` has been accepted by the system, but waiting for one or more of it's pre-requisites to be completed. Such as waiting for required [TLS certificates](../README.md#security) to be generated |
+| Initializing | The required sub-resources of the `Deployment` are being initialized |
+| Running | All the required sub-resources of the `Deployment` are created. This does not mean that the deployed driver is fully ready to use, there might still possibility of failures. Reasons for such failures can be determined by looking into individual pod status |
+| Failed | For some reason the state of the `Deployment` failed and cannot be progressed<sup>2</sup> |
+
+<sup>2</sup> Failure reason is supposed to be carried by one of additional `DeploymentStatus` filed, but not implemented yet.
+
+### TLS Certificate approval
+
+All [PMEM-CSI driver communication is protected by mutual TLS](../README.md#security). So the driver
+deployment needs to generate those certificates. The operator deals this in 
+the form of Kubernetes [CertificateSigningRequest](https://kubernetes.io/docs/tasks/tls/managing-tls-in-a-cluster/) objects. On a new deployment,
+the operator creates required `CertificateSigningRequest` Kubernetes objects
+and keeps the driver deployment in `pending` state till those CSRs get approved by 
+an authorized user or component. Once the CSRs are approved and certificates 
+are ready, then the deployment process proceeds.
+
+By default the operator runs a dedicated controller for auto-approving these CSR
+requests. This controller uses Kubernetes Authority (CA) provided by the Kubernetes
+cluster to get those certificates. The controller watches for new CSRs and approves 
+them by ensuring that those requests are generated by a valid deployment and the 
+request subject matches to a valid _CommonName(CN)_ expected by PMEM-CSI driver.
```

Comment:
This explains *what* is done, but not *why*. Why are we using mutual TLS (first sentence in this section) and why is that goal still reached when using this approach explained here?

The way it's explained here it makes it sound like our security model is intact because the auto-approver is careful about what it signs. But after thinking a bit about this, I believe the opposite is true: because our approver is careful to not sign unrelated CSRs, the security of *other* components in the system is preserved when enabling it.

We can't protect ourselves against a malicious or careless approver. If they approve a CSR for "our" names although it wasn't the operator which created the CSR, then valid certificates will be created that allow other entities to communicate with PMEM-CSI. But that's okay: the cluster admin has to take care that no-one can approve CSRs without due diligence.

Does that make sense? If yes, can you update the text here?

Reply:
> This explains _what_ is done, but not _why_. Why are we using mutual TLS (first sentence in this section) and why is that goal still reached when using this approach explained here?

Do you mean this section should also explain 'why are we using mutual TLS' and/or 'why we choose this approving model'?

Reply:
@pohly, Does this sounds any better:

```
By default, the operator runs a dedicated controller for auto-approving these
CSR requests without any manual intervention. This controller uses the
Certificate Authority (CA) provided by the Kubernetes cluster to get those
certificates. The controller watches for new CSRs and approves them by ensuring
that those requests are generated by a valid deployment and the request subject
matches a valid _CommonName(CN)_ expected by PMEM-CSI driver. This way we ensure
that the controller is careful enough in not approving any CSRs in the system
other than the ones requested by PMEM-CSI deployment(s).
```

Reply:
This proposal looks good. Let me add to it a bit.

> Do you mean this section should also explain 'why are we using mutual TLS'

As this section mentions it, yes. Something like:

> All [PMEM-CSI driver communication is protected by mutual TLS](../README.md#security) to ensure that malicious or compromised entities in the cluster cannot interfere with its operation.
> ...
> This way we ensure
that the controller is careful enough in not approving any CSRs in the system
other than the ones requested by PMEM-CSI deployment(s). 
> The assumption is that other controllers and/or the cluster admins are equally careful to not sign CSRs for the special names used by PMEM-CSI. Under that assumption, only PMEM-CSI is going to have valid certificates for those names and thus its internal communication is protected.

Reply:
Ok, Thanks. I pushed these changes.

Reply:
I'm having second thoughts about this:
> The way it's explained here it makes it sound like our security model is intact because the auto-approver is careful about what it signs. But after thinking a bit about this, I believe the opposite is true: because our approver is careful to not sign unrelated CSRs, the security of other components in the system is preserved when enabling it.

This is not the full truth. In addition to protecting others, we *also* need to protect ourselves from an attacker which makes a CSR look like it was created by us.

Does the current solution achieve that reliably? That hinges on how it determines what "a valid deployment" is. Can you elaborate on that?

Reply:
> Does the current solution achieve that reliably? That hinges on how it determines what "a valid deployment" is. Can you elaborate on that?

By checking the if the owner-reference of the CSR is a ' `Deployment('pmem-csi.intel.com/v1alph)` and if that deployment is really in `pending` state that means waiting for csr approval.

Reply:
That still doesn't answer why an attacker can't create such a CSR.

There are multiple aspects to this:
- If the attacker is able to create such a deployment, then he's by definition allowed to set up a deployment and thus has access to the resulting secrets as far as PMEM-CSI is concerned. Cluster admins thus must control who gets this permission.
- If the attacks is unable to create such a deployment, he might still try to create a CSR while some other deployment is pending. This is a race, but we shouldn't rely on timing for security. What saves us here is that ownership is only possible for CSRs in the same namespace. Therefore an attacker must not get permission to create CSRs in the namespace of a deployment.

We need to document this as guidance for admins and explain in the discussion of "why this is secure" that we are making these assumptions.

Reply:
> and if that deployment is really in pending state that means waiting for csr approval.

That isn't relevant, is it? It doesn't add any additional defense against an attacker.

Reply:
> * What saves us here is that ownership is only possible for CSRs in the same namespace. Therefore an attacker must not get permission to create CSRs in the namespace of a deployment.

CSRs are cluster level resources, not namespace scoped :-(

Reply:
> CSRs are cluster level resources, not namespace scoped :-(

And the PMEM-CSI deployment then also is not namespaced? I thought it was because it's not mentioned that it isn't and the example output for `kubectl describe deployment.pmem-csi.intel.com pmem-deployment` shows a `namespace: default`. Or is that always shows, regardless whether the CSR is namespaced?

But apparently they are not, because otherwise you shouldn't be able to set the deployment as owner of the non-namespaced CSR. https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#owners-and-dependents says "2) Cluster-scoped dependents can only specify cluster-scoped owners, but not namespace-scoped owners."

Let's take a step back. Why are we using CSRs at all? I can only think of one reason, which is "then someone might theoretically write a custom approver". We have no idea whether that is even possible in a secure way. What about just dropping this approach entirely for now until we actually have someone asking for it, and instead do CA creation and certificate signing in the operator itself?

What was the rationale for using signing certificates? They were introduced in https://github.com/intel/pmem-csi/commit/5b6432c6da44a339de3e58f14760b1f60d067564#diff-ea37666df77a82a46982984778217075 but that commit message only says what is in the commit, not why it is done this way.

Reply:
> And the PMEM-CSI deployment then also is not namespaced? I thought it was because it's not mentioned that it isn't and the example output for `kubectl describe deployment.pmem-csi.intel.com pmem-deployment` shows a `namespace: default`. Or is that always shows, regardless whether the CSR is namespaced?
>

No,  PMEM-CSI deployments are namespace scoped. [This](https://github.com/intel/pmem-csi/pull/525/commits/4f21a630624dbc2053b8effb0d356d7ef8031591) change adds support for deploying in a different namespace other than the operator's namespace.

> But apparently they are not, because otherwise you shouldn't be able to set the deployment as owner of the non-namespaced CSR. https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#owners-and-dependents says "2) Cluster-scoped dependents can only specify cluster-scoped owners, but not namespace-scoped owners."

Hmm, This is bit unclear to me, where this restriction is imposed. But, somehow it worked for me with no issues. I was setting owner reference of CSR(in fact other cluster-scoped objects created by deployment - ClusterRole, ClusterRoleBinding) with deployment information. Initially, i was using for cleaning-up of these objects whenever deployment gets deleted.

Reply:
> Let's take a step back. Why are we using CSRs at all?

The idea behind using CSRs are for utilizing the Kubernetes's certificates subsystem and CA.
Kubernetes service account CA certificate is available to all Pods so that we need to create our own CA and deploy CA certificate to Pod.
 
 > I can only think of one reason, which is "then someone might theoretically write a custom approver". 

This is the reason for dividing CSR creation and approval as separate components.

>  We have no idea whether that is even possible in a secure way. 

If this is the case, then we can avoid the certificate approver controller and the operator itself creates an approved CSR, instead of waiting for someone to approve?.

Reply:
> Hmm, This is bit unclear to me, where this restriction is imposed.

You can only set a `Name` for the owner, not `Namespace`: https://godoc.org/github.com/ericchiang/k8s/apis/meta/v1#OwnerReference

You can set a UUID. If that works across namespaces, then the Kubernetes documentation should be updated to reflect that.

> But, somehow it worked for me with no issues. I was setting owner reference of CSR(in fact other cluster-scoped objects created by deployment - ClusterRole, ClusterRoleBinding) with deployment information.

With UUID or Name? Probably the former, otherwise you would have noticed the lack of the `Namespace` field. Have you tested that this works in a non-default namespace?

> Kubernetes service account CA certificate is available to all Pods so that we need to create our own CA and deploy CA certificate to Pod.

Which we can do. We can simply put it into the same secret as the client certificate.

> If this is the case, then we can avoid the certificate approver controller and the operator itself creates an approved CSR, instead of waiting for someone to approve?.

Correct.

Reply:
> With UUID or Name? Probably the former, otherwise you would have noticed the lack of the Namespace field. Have you tested that this works in a non-default namespace?

In case it wasn't obvious: with "that this works" I meant "garbage collection deletes orphaned objects", not just "apiserver allows to set ownership".

Reply:
> In case it wasn't obvious: with "that this works" I meant "garbage collection deletes orphaned objects", not just "apiserver allows to set ownership".

Yes, delete deploy resulted in the cleaning of all its sub-resources owned including cluster-level objects.

Reply:
> Have you tested that this works in a non-default namespace?

Yes!

Reply:
It worked only by accident: https://github.com/kubernetes/kubernetes/issues/65200

Reply:
Do we want the PMEM-CSI deployment to be namespaced? It's a privileged object, only admins should be allowed to create it. Given that it creates some non-namespaced objects, I believe it makes more sense to make it non-namespaced.

Reply:
> It worked only by accident: [kubernetes/kubernetes#65200](https://github.com/kubernetes/kubernetes/issues/65200)

:-)

Reply:
> Do we want the PMEM-CSI deployment to be namespaced? It's a privileged object, only admins should be allowed to create it. Given that it creates some non-namespaced objects, I believe it makes more sense to make it non-namespaced.

Ok. I have no strict opinion on this. In fact, those cluster-level objects need not to betide to a deployment. We could pre-create them and share them with all deployment objects.

If you believe it should be cluster scoped, I could implement those changes.

Reply:
> Those cluster-level objects need not to betide to a deployment. We could pre-create them and share them with all deployment objects.

Do we really have objects which can be reused by different deployments? Which ones? If we have them and can't create them per deployment, then that sounds like a problem for cleanup (would need to track number of deployments that still need them).

> If you believe it should be cluster scoped, I could implement those changes.

Ownership relationships are useful and we can't do them when it's namespaced, so yes, I think this should be changed.

Reply:
> >   It worked only by accident: kubernetes/kubernetes#65200
>
> :-)

Lesson learned: don't use experiments to figure out how an API works, follow the documentation...